File: Chapter 09/C#/Section_9_2_3_Stateful_Vs_Stateless_Design/Services/UserService.cs
namespace Chapter09.StatefulVsStateless.Services;

/// <summary>
/// THE APPLICATION LAYER (Core Business Logic):
/// 
/// TEACHING NOTE:
/// Look at the 'using' statements at the top of this file. There is no 'using System.IO;' 
/// and no AWS SDK imports. This class is blissfully ignorant of where the files actually go.
/// By keeping infrastructure out of our domain, this class becomes incredibly easy to test 
/// and completely cloud-agnostic.
/// </summary>
public class UserService
{
    private readonly IStorageProvider _storage;

    // We inject the dependency (The Adapter) through the constructor.
    // The UserService doesn't build its own database; it asks for one to be provided.
    public UserService(IStorageProvider storage)
    {
        _storage = storage;
    }

    public void UploadAvatar(string userId, string imageData)
    {
        string fileName = $"profile_{userId}.jpg";
        _storage.Save(fileName, imageData);
    }

    public string ViewAvatar(string userId)
    {
        string fileName = $"profile_{userId}.jpg";
        return _storage.Get(fileName);
    }
}

File: Chapter 09/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter09/section_9_2_3_stateful_vs_stateless_design/services/UserService.java
package com.grokkingsoftwarearchitecture.chapter09.section_9_2_3_stateful_vs_stateless_design.services;

/**
 * THE APPLICATION LAYER (Core Business Logic):
 * * TEACHING NOTE:
 * Look at the imports at the top of this file. There is no 'java.io.File' 
 * and no 'com.amazonaws.services.s3'. This class is blissfully ignorant of where 
 * the files actually go. By keeping infrastructure out of our domain, this class 
 * becomes incredibly easy to test and completely cloud-agnostic.
 */
public class UserService {
    
    private final StorageProvider storage;

    // We inject the dependency (The Adapter) through the constructor.
    // The UserService doesn't build its own database/file system; it asks for one to be provided.
    public UserService(StorageProvider storage) {
        this.storage = storage;
    }

    public void uploadAvatar(String userId, String imageData) throws Exception {
        String fileName = "profile_" + userId + ".jpg";
        this.storage.save(fileName, imageData);
    }

    public String viewAvatar(String userId) throws Exception {
        String fileName = "profile_" + userId + ".jpg";
        return this.storage.get(fileName);
    }
}